Loader modularization

dylandreimerink opened the pull request:
This PR turns the eBPF loader in `pkg/datapath/loader` into a Hive cell. It also unexported all symbols where possible. Global functions which were called outside of the package have been moved to the Loader interface so we have a single clear API boundary, making it easier to improve upon in the future.

This step will also make it easier to move non-loader logic outside of the package by allowing us to then still depend on the logic via hive.

Part of #30024

dylandreimerink committed:
loader: Make loader API boundary more explicit

This commit unexports all types and functions on the package itself
except for the Loader interface. All functions accessed from outside
the package now are methods of this interface.

This makes the API boundary of the loader package more explicit and
easier to reason about and refactor. It is also a first step towards
turning the loader package into a module.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

rgo3 reviewed (approved):
LGTM, pulled in Timo as well to have more loader :eyes: on this change.

asauber reviewed (approved):
cilium-dbg related changes lgtm

ti-mo commented on pkg/datapath/loader/netlink.go:
```diff
@@ -391,7 +391,7 @@ func setupVethPair(name, peerName string) error {
 // creation, if needed) of those interfaces. It returns two links and an error.
 // By default, it sets up the veth pair - cilium_host and cilium_net.
-func SetupBaseDevice(mtu int) (netlink.Link, netlink.Link, error) {
+func (l *loader) SetupBaseDevice(mtu int) (netlink.Link, netlink.Link, error) {
```
We can avoid leaving this exported and adding this to the loader API at all by making `createDevices()` in `daemon_privileged_tests.go` create a dummy device named `defaults.HostDevice`. All other uses are internal, and for the sake of the privileged tests, any device will do. cilium_net isn't even touched.

ti-mo commented on pkg/datapath/loader/template.go:
```diff
@@ -282,6 +282,6 @@ func elfVariableSubstitutions(ep datapath.Endpoint) map[string]uint64 {
 // must be implemented against an ELF template to configure the datapath for
 // the specified endpoint.
-func ELFSubstitutions(ep datapath.Endpoint) (map[string]uint64, map[string]string) {
+func (l *loader) ELFSubstitutions(ep datapath.Endpoint) (map[string]uint64, map[string]string) {
```
Nit: can this and its dependent functions be moved to package endpoint?

dylandreimerink replied:
Actually, its only used internally, except for in the tests of `pkg/datapath/linux/config`. And I don't see a what to not brick the tests. We should merge the header file writing logic of `pkg/datapath/linux/config` into `pkg/datapath/loader` at which point we can make it an internal function

ti-mo reviewed (approved):
Thanks! Left a few suggestions, but these can be picked up in follow-ups as well.

dylandreimerink committed:
loader: Make loader into cell and provide to hive

This commit changes the loader into a cell. The Datapath object depends
on it and re-distributes it still (for now). All locations that created
a loader now reuse the loader instance from hive. The exception being
tests which did not yet use hive.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

mhofstetter reviewed (approved):
`sig-agent` :heavy_check_mark: 

Nice - thanks Dylan!
